### ISSUE TITLE
A problem with loss computation.

### ISSUE THREAD

**Issue by yxdr** [bentrevett/pytorch-seq2seq#66]
output = output[1:].view(-1, output_dim)
trg = trg[1:].view(-1)
loss = criterion(output, trg)

Above is a part of your codes. The loss computed by the above line is the average at every time step, which can cause it difficult to train the model.
So I suggest accumulating the loss at every time step. In my experiments, this makes it easier to train the model.

**Comment by bentrevett** [comment]
@yxdr

I've experimented with this before and found that you are correct, it does give lower loss values. I'll double check it actually gives better translations/better BLEU scores and then make it from a per sequence loss instead of per token.

Do you have any intuition of why training this way gives a lower loss? Can't think of one off the top of my head.

**Comment by thak123** [comment]
@bentrevett  its because the gradients are fluctuating from - to + ve extremes like sine curve for every update that the network does for each of the time stamp. If one accumulates all the gradient for one sequence then the updated values of the weights/parameters would be the average of the gradients  faster to converge in less number of epochs.  I read this on one article from this repository https://github.com/allenai/writing-code-for-nlp-research-emnlp2018 . I may be wrong but this could be the answer.

**Comment by bentrevett** [comment]
@thak123 Thanks for the link, I'll check it out.

**Comment by yxdr** [comment]
> @yxdr
> 
> I've experimented with this before and found that you are correct, it does give lower loss values. I'll double check it actually gives better translations/better BLEU scores and then make it from a per sequence loss instead of per token.
> 
> Do you have any intuition of why training this way gives a lower loss? Can't think of one off the top of my head.

Accumulating the loss during decoding makes the total loss used to compute gradients t times larger, where t is the length of the target sequences, and so makes the gradients larger.  If you use the reshaped trg and output to compute the total loss, the total loss will be smaller. 
This is my intuition, but I am not very sure that this is right.

**Comment by bentrevett** [comment]
> > @yxdr
> > I've experimented with this before and found that you are correct, it does give lower loss values. I'll double check it actually gives better translations/better BLEU scores and then make it from a per sequence loss instead of per token.
> > Do you have any intuition of why training this way gives a lower loss? Can't think of one off the top of my head.
> 
> Accumulating the loss during decoding makes the total loss used to compute gradients t times larger, where t is the length of the target sequences, and so makes the gradients larger. If you use the reshaped trg and output to compute the total loss, the total loss will be smaller.
> This is my intuition, but I am not very sure that this is right.

That makes sense. I think I assumed that the larger gradients would give a more unstable learning experience. If the gradients just need to be larger, could we not just increase the learning rate?

**Comment by thak123** [comment]
@bentrevett idea is not to have large gradients,as it will cause unstable learning. the aim is to update in such a way that bouncy gradient values are not attained (extreme -ve and +ves)

**Comment by kushalj001** [comment]
@thak123 Can you tell how this can be fixed in code?

**Comment by bentrevett** [comment]
To set it so the losses are summed over the sequences and then averaged, thus giving you average loss per sequence instead of loss per sequence, first change the criterion to:

```python
criterion = nn.CrossEntropyLoss(ignore_index = TRG_PAD_IDX, reduction = 'sum')
```

That sums all of the losses for every single token. You then want to divide this by the batch size to get average loss per sequence, which can be done when we calculate the loss with the criterion as:

```python
loss = criterion(output, trg).div_(batch_size)
```

**Comment by bentrevett** [comment]
So I have now done quite a few experiments using the per sentence loss instead of per token.

For these models and the dataset used for these tutorials the per token loss gives the superior BLEU quality, usually around 5 BLEU points higher.

Therefore I am going to keep the loss calculations as they are and close this issue.
